src/ray/raylet/worker_pool.cc

```diff
@@ -894,7 +863,12 @@ std::shared_ptr<WorkerInterface> WorkerPool::PopWorker(
       // There are no more non-actor workers available to execute this task.
       // Start a new worker process.
       proc = StartWorkerProcess(task_spec.GetLanguage(), rpc::WorkerType::WORKER,
```

Comment:
One issue: StartWorkerProcess could be called multiple times if PopWorker is called again before the new worker has registered and been pushed into idle_of_all_languages in PushWorker.

Reply:
Another issue is a race condition.  The new worker is placed into idle_of_all_workers when it's ready, but it may be killed by TryKillingIdleWorkers before PopWorker is called again, so it won't get a chance to run its task, and PopWorker will start a new worker process here again.

Reply:
Is this issue from the previous code or happening due to your changes?

Reply:
I spoke with @edoakes about these.  For the first issue, I think it is present in the previous code, but it didn't seem to be causing problems--maybe something to do with the scheduler limiting calls to PopWorker based on resources, but I'm not sure.  Through debug print statements I've seen PopWorker get called 2-3 times before the new worker is registered.  It could potentially be more of an issue in the future when we have worker processes that take several minutes to start--for example, if a runtime env conda environment needs to be installed before the worker can start.

The second issue would also be present in the previous code, but it also didn't seem to be causing problems.  It's probably rare or impossible (Idle workers are killed in FIFO order, so all the other workers would have to be killed first).  Also, even if it does happen, it will just start a new worker process, so it's probably not an issue.

Reply:
hmm I see. You might be right issues are not revealed because worker process startup time is short now. Maybe once we discover issues after we start testing runtime env workers, we can start fixing them .